heron/spi/src/java/com/twitter/heron/spi/statemgr/IStateManager.java

```diff
@@ -276,4 +277,27 @@ public String getName() {
    */
   ListenableFuture<PhysicalPlans.PhysicalPlan> getPhysicalPlan(
       WatchCallback watcher, String topologyName);
+
+  /**
+   * Set the Stateful Checkpoint
+   *
+   * @return Boolean - Success or Failure
+   */
+  ListenableFuture<Boolean> setStatefulCheckpoint(
+      CheckpointManager.StatefulConsistentCheckpoints checkpoint, String topologyName);
```

Comment:
Storing a list of checkpoints achieves the same result as the solution you proposed here. Plus it's much more simpler in terms of implementation and management.

To manage those checkpoints, the TMaster could fetch all of them in the state managers and then modifies the fetched states and then write it back.